src/api/jsonize_attr_string.cpp

```diff
@@ -0,0 +1,237 @@
+#include "jsonize_attr_string.hpp"
+#include <vector>
+#include <stack>
+#include <iostream>
+#include <algorithm>
+#include <cassert>
+#include <unordered_map>
+#include <unordered_set>
+#include <utility>
+#include <migraphx/errors.hpp>
+
+namespace migraphx {
+inline namespace MIGRAPHX_INLINE_NS {
+
+// get all elements of an array or an object, including '[]' and '{}'
+std::string get_elements_string(const std::string& str, const std::size_t start, const char brkt)
```

Comment:
We dont need to parse the json. We can just lex the json(and skip whitespace). When the next token is `:` then we can check if the previous token has quotes, if not then we add quotes while we serialize it to a string.

Each token can be an iterator(or pointer) pair so we dont have to make copies of the string as we lex.

Reply:
Actually you could use something like this to tokenize the json string:

```cpp

template<class Predicate>
auto lex_while(P p)
{
    return [=](const char* start, const char* end) {
        return std::find_if(start, end, [&](char c) {
            return not p(c);
        });
    };
}

template<class Predicate>
auto lex_if(P p)
{
    return [=](const char* start, const char* end) {
        if (p(*start))
            return start+1;
        return start;
    };
}

using token = std::pair<const char*, const char*>;
using lexer = std::function<const char*(const char* start, const char* end)>;

std::vector<token> tokenize(const char* start, const char* end, std::vector<lexer> lexers)
{
    std::vector<token> result;
    while(start != end) 
    {
        bool error = true;
        for(auto l:lexers) 
        {
            auto next = l(start, end);
            if (next != start)
            {
                if (not std::all_of(start, next, &isspace))
                    result.emplace_back(start, next);
                start = next;
                error = false;
                break;
            }
        }
        if (error)
            MIGRAPHX_THROW("Lex error");
    }
    return result;
}

std::vector<token> json_tokenize(const std::string& s)
{
    std::vector<lexer> lexers;
    // Quote
    lexers.push_back([](const char* start, const char* end) {
        if (*start != '\"')
            return start;
        while(start != end and *start != '\"')
        {
            if (*start == '\\')
                start++;
            start++;
        }
    });
    // Whitespace
    lexers.push_back(lex_while(&isspace));
    // Punctation
    lexers.push_back(lex_if(&ispunct));
    // Identifier/number
    lexers.push_back(lex_while([](char c) { return isalnum(c) or '_' or '.' or '+'; }));
    return tokenize(s.data(), s.data()+s.length(), lexers);
}
```

As we lex identifiers, we can add quotes to them if it doesn't start with a number, which should add quotes for keys or values. You could also extend the lexer to tokenize floating point and integers separately.